Add AuxiliaryClassFiles in IncOptions

adpi2 opened the pull request:
Fixes #579 

Add an `AuxiliaryClassFiles` member in `IncOptions` so that the `ClassFileManager` manages the files that are associated to class files. This will be used for Scala 3 (`.tasty` files), for Scala.js (`.sjsir` files) and for Scala Native (`.nir` files).

The `auxiliaryClassFiles` member in `IncOptions` is an array so that we can manage several type of files at the same time (for instance `.tasty` and `.sjsir` files). Its default value is an empty array because Scala 2 and Java don't need any.

The implementation for `.tasty` files is provided so that it can already be reused by sbt (https://github.com/sbt/sbt/issues/6080) and also by Mill and Bloop.

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=develop&zinc=pull/956/head&io=develop&librarymanagement=develop&util=develop&website=develop&deploy_to_bintray=false&run_sbt_community=false&verbose_dbuild=false&community_verbose=false">this link.</A>

eed3si9n commented:
Could you exercise this from https://github.com/sbt/zinc/blob/develop/zinc/src/test/scala/sbt/inc/IncrementalCompilerSpec.scala or scripted test somehow?

smarter commented:
Thanks for working on this!

> The auxiliaryClassFiles member in IncOptions is an array so that we can manage several type of files at the same time (for instance .tasty and .sjsir files). Its default value is an empty array because Scala 2 and Java don't need any.
> The implementation for .tasty files is provided so that it can already be reused by sbt (sbt/sbt#6080) and also by Mill and Bloop.

I  wonder if besides .tasty we shouldn't also hardcode .sjsir and .nir by default, it shouldn't break anything and it means that mill/bloop/maven/... don't have to remember to do this themselves. /cc @sjrd

sjrd commented on internal/compiler-interface/src/main/java/xsbti/compile/AuxiliaryClassFiles.java:
```diff
@@ -0,0 +1,14 @@
+import java.nio.file.Path;
+
+/**
+ * Interface that associate a class file with other corresponding produced
+ * files, for instance `.tasty` files or `.jsir` files.
```
```suggestion
 * Interface that associates a class file with other corresponding produced
 * files, for instance `.tasty` files or `.sjsir` files.
```

sjrd commented on internal/zinc-core/src/main/java/xsbti/compile/TastyFiles.java:
```diff
@@ -0,0 +1,23 @@
+    @Override
+    public Path[] associatedFiles(Path classFile) {
+        if (classFile.endsWith(".class")) {
+            String absolutePath = classFile.toAbsolutePath().toString();
+            String prefix = absolutePath.substring(0, absolutePath.length() - 6);
+            Path tastyFile = Paths.get(URI.create(prefix + ".tasty"));
+            return new Path[] { tastyFile };
```
This implementation indiscriminately returns a `Path` for a `.tasty` file even if it doesn't exist. Not all `.class` files have a corresponding `.tasty` file (only top-level classes have one). Is this expected, i.e., is the client supposed to ignored returned `Path`s that do not exist?

adpi2 replied:
Yes I think it is easier if the `ClassFileManager` is forgiving on auxiliary files that don't exist.
Also, as far as I can see, the `ClassFileManager` implementations are already forgiving on class files that don't exist.

I add a comment and a test case.

sjrd commented:
> I wonder if besides .tasty we shouldn't also hardcode .sjsir and .nir by default, it shouldn't break anything and it means that mill/bloop/maven/... don't have to remember to do this themselves.

We could do that. I don't have a strong opinion either way. If Zinc doesn't do it by default, sbt-scalajs and sbt-scala-native will keep doing it.

adpi2 committed:
Add AuxiliaryClassFiles

Add auxiliaryClassFiles member in IncOptions to manage files that are
associated to class files. Provide an implementation for TASTy files.

adpi2 committed:
Add TastyFiles ScalaJSFiles and ScalaNativeFiles

adpi2 committed:
Test AuxiliaryClassFiles

sjrd commented on zinc/src/test/scala/sbt/inc/AuxiliaryClassFilesSpec.scala:
```diff
@@ -0,0 +1,29 @@
+import java.nio.file.Path
+
+class AuxiliaryClassFilesSpec extends BaseCompilerSpec {
+  it should "allow client to add define their own auxiliary class files" in {
```
```suggestion
  it should "allow client to define their own auxiliary class files" in {
```

adpi2 replied:
this test does not exist anymore because it was useless

sjrd commented on internal/zinc-core/src/main/java/xsbti/compile/TastyFiles.java:
```diff
@@ -0,0 +1,25 @@
+    public Path[] associatedFiles(Path classFile) {
+        // not all classfiles have a corresponding TASTy file but we strongly assume that the
+        // ClassFileManager is forgiving on files that do not exist.
+        String absolutePath = classFile.toAbsolutePath().toString();
+        if (absolutePath.endsWith(".class")) {
+            String prefix = absolutePath.substring(0, absolutePath.length() - 6);
+            Path tastyFile = Paths.get(prefix + ".tasty");
```
This implementation will only work for `Path`s that represent the real file system. It would be better, IMO, to only use methods of `Path` that work for arbitrary `Path`s. Perhaps the following:
```suggestion
        String fileNameString = classFile.getFileName().toString();
        if (fileNameString.endsWith(".class")) {
            String fileNamePrefix = fileNameString.substring(0, fileNameString.length() - 6);
            Path tastyFile = classFile.resolveSibling(fileNamePrefix + ".tasty");
```

sjrd replied:
The same comment applies to the other implementations of `AuxiliaryClassFiles`, obviously.

adpi2 replied:
I have applied those changes and reduced the code duplication by introducing a `AuxiliaryClassFileExtension` parent class

sjrd commented on internal/zinc-core/src/main/java/xsbti/compile/ScalaJsFiles.java:
```diff
@@ -0,0 +1,22 @@
+ * Can be added to `IncOptions.auxiliaryClassFiles` so that the sjsir files
+ * produced by the Scala.js compiler are managed by the ClassFileManager.
+ */
+public class ScalaJsFiles implements AuxiliaryClassFiles {
```
I don't know what's the convention of this repo for acronyms in class names, but in Scala.js we typically `ScalaJS`, for example `ScalaJSPlugin`. Under that convention, it would be nice for this to be called `ScalaJSFiles`. However, that may conflict with the conventions of this repo, which should prevail.

sjrd reviewed (approved):
Overall LGTM. I have a few comments.
